src/main/java/com/conveyal/datatools/manager/UpdateSQLFeedsMain.java

```
@@ -1,131 +1,107 @@
-            String namespace = resultSet.getString(1);
-            String updateTableSql = updateSql.replaceAll(NAMESPACE_WILDCARD, namespace);
-            Statement statement = connection.createStatement();
-            try {
-                System.out.println(updateTableSql);
-                int updated = statement.executeUpdate(updateTableSql);
-                System.out.println(String.format("Updated rows: %d", updated));
-                successCount++;
-            } catch (SQLException e) {
-                // The stops table likely did not exist for the schema.
-                e.printStackTrace();
-                failedNamespace.add(namespace);
+
+            SqlSchemaUpdater schemaUpdater = new SqlSchemaUpdater(connection);
+            Collection<NamespaceCheck> checkedNamespaces = schemaUpdater.checkReferencedNamespaces(namespaceTypesToCheck);
+
+            for (NamespaceCheck namespaceCheck : checkedNamespaces) {
+                String namespace = namespaceCheck.namespace;
+                try {
+                    if (!namespaceCheck.isOrphan()) {
+                        schemaUpdater.upgradeNamespaceIfNotOrphan(namespaceCheck);
+                        System.out.printf("Updated namespace %s%n", namespace);
+                        successCount++;
+                    }
+                } catch (Exception e) {
+                    e.printStackTrace();
+                    failedNamespaces.add(namespace);
+                }
+            }
+
+            System.out.println(String.format("Updated %d namespaces.", successCount));
+            System.out.printf("Failed namespaces (%d):%n%s%n", failedNamespaces.size(), String.join("\n", failedNamespaces));
+            // No need to commit the transaction because of auto-commit above (in fact, "manually" committing is not
+            // permitted with auto-commit enabled.
+            if (testRun) {
```

I think this should be in a`finally` block so it guaranteed to be run in cases where an SQL exception is throw by `SqlSchemaUpdater`.

---

Good point. Should we rollback everything on the first exception (i.e. if any namespace fails)?

---

On second review no alterations have been made to any namespace at this point:

`Collection<NamespaceCheck> checkedNamespaces = schemaUpdater.checkReferencedNamespaces(namespaceTypesToCheck);`

All namespace updates (inside the for loop) are wrapped in a catch all exception. This is going to catch all SQL exceptions and prevent them from rise up the stack. Therefore, I think what is there will be ok and the rollback will be executed.